### PR TITLE
make LevelGroup use child_levels relationship with its sublevels

### PR DESCRIPTION
**Pull request by davidsbailey**
## Background

See https://github.com/code-dot-org/code-dot-org/pull/34677

## Description

1. move LevelGroup "levels" and "texts" out of level properties and into the parent_levels_child_levels table
2. as a result of (1), we can no longer have the same "text" appear twice in the same level group. 5431a5a2162293f33662c6d736fa4d12316372c7 fixes the only instance of this.

The following comment best explains the current terminology:
https://github.com/code-dot-org/code-dot-org/blob/e3d49ef3d8a2d80e13f0a9378d3cd45497acba24/dashboard/app/models/levels/level_group.rb#L85-L101

## Testing story
- [x] existing unit and UI tests are passing
- [x] added some unit test coverage for new implementation of texts and page internals
- [x] added ui test coverage for level numbers
- [x] manually verified texts are appearing in the right place throughout https://studio.code.org/s/csp-post-survey/stage/1/puzzle/1

## Future work

I chose to keep `page.levels` named the same and added `levels_and_texts` to maintain existing terminology and avoid lots of renames. However, the terminology is confusing because "levels" and "texts" both represent Level objects. If there is buy-in, here is one way we could improve the terminology" in a next PR:

| old name | new name |
|---|---|
| levels | questions |
| texts | texts |
| levels and texts | levels |

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by davidsbailey** [comment]
> ## Future work
> I chose to keep `page.levels` named the same and added `levels_and_texts` to maintain existing terminology and avoid lots of renames. However, the terminology is confusing because "levels" and "texts" both represent Level objects. If there is buy-in, here is one way we could improve the terminology" in a next PR:
> 
(see above)

@clareconstantine and @jmkulwik , just wanted to include you in case you have input on the naming here, since these are terms LP will work with here. Happy to provide additional context.

File: dashboard/app/dsl/level_group_dsl.rb [hunk 1]
```diff
@@ -39,6 +38,12 @@ def page
   end
 
   def text(name)
+    # Ensure level name hasn't already been used.
+    if @level_names.include? name
+      raise "Don't use the same level twice in a LevelGroup (#{name})."
+    end
+    @level_names << name
```
Comment (dmcavoy) [inline dashboard/app/dsl/level_group_dsl.rb]: Is this really now @level_and_text_names?

File: dashboard/app/dsl/level_group_dsl.rb [hunk 2]
```diff
@@ -49,9 +54,7 @@ def text(name)
       raise "LevelGroup text must be a level of type external. (#{name})"
     end
 
-    # Record the name of the external level, as well as the index of the actual level
-    # it appears immediately before.
-    @hash[:texts] << {level_name: name, index: @level_names.length}
+    @current_page_level_names << name
```
Comment (dmcavoy) [inline dashboard/app/dsl/level_group_dsl.rb]: is this really current_page_level_and_text_names?

**Review by dmcavoy — APPROVED** [review]
Looks good! I left a couple questions about naming in level_group_dsl but nothing blocking.

**Review by uponthesun — COMMENTED** [review]
I'm not planning to review this in-depth since I don't have much context on LevelGroups, but changes LGTM at a glance.

**Pull request merged**